### DISCUSSION TITLE
support for ASN.1 value notation?

### DISCUSSION THREAD

**Discussion by laf0rge** [librasn/rasn#216] [Q&A]
I couldn't find any reference regarding the support for ASN.1 value notation.  To the human, it looks rather similar to the textual representation of JSER or GSER.  However, it is syntactically different.

There are standards/interfaces where data is exchanged in ASN.1 value notation, and hence it is needed to encode/decode that notation.  One can look at it as yet another set of encoding rules.

References: 
* https://www.oss.com/asn1/knowledge-center/asn1-tools-general/value-notation.html
* https://stackoverflow.com/questions/56453965/how-to-parse-asn-1-value-notation-format

Is this something that rasn supports or plans to support? Thanks in advance!

**Comment by Nicceboy** [comment]
I think this is something what you would need to write yourself in Rust. Value notation here is to program ASN.1 as Rust. 

E.g. from the second link
```
value Rocket ::= {
  name "Falcon",
  -- use default for the message --
  fuel solid,
  speed mph : 18000,
  payload {
    "Car",
    "GPS"
  }
}
```

Simplified assumed sample objects and corresponding usage:

```rust 
#[derive(AsnType, Decode, Encode, Clone, Copy, Debug, PartialEq)]
#[rasn(enumerated, automatic_tags)]
pub enum Fuel {
    Solid,
    Liquid,
    Gas,
}

#[derive(AsnType, Decode, Encode, Clone, Debug, PartialEq)]
#[rasn(automatic_tags)]
pub struct Speed {
    pub mph: Integer,
}

#[derive(AsnType, Decode, Encode, Clone, Debug, PartialEq)]
#[rasn(automatic_tags)]
pub struct Rocket {
    pub name: Utf8String,
    pub fuel: Fuel,
    pub speed: Speed,
    pub payload: Vec<Utf8String>,
}

let rocket = Rocket {
    name: "Falcon".to_string(),
    fuel: Fuel::Solid,
    speed: Speed { mph: 18000 },
    payload: vec!["Car".to_string(), "GPS".to_string()],
};
```
Now you can encode `rocket` values to any `rasn` supported codec values.


Check this PR for more examples:

https://github.com/XAMPPRocky/rasn/pull/159

**Reply by XAMPPRocky** [reply]
@Nicceboy I think they mean adding a codec that encodes the code you wrote in the second example, into the code in your first example.

**Reply by Nicceboy** [reply]
Oh I see, oops.

**Comment by XAMPPRocky** [comment]
Thank you for question, there isn't support right now, but I'm not opposed to accepting a contribution adding support for encoding values into that notation.
